compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ParameterizeGlobals.cpp

```diff
@@ -0,0 +1,248 @@
+      (iree_host_size_t)archive_length, host_allocator, &file_contents));
+  iree_io_file_handle_release_callback_t release_callback = {
+      iree_io_file_handle_release_mapping, file_contents};
+  iree_status_t status = iree_io_file_handle_wrap_host_allocation(
+      IREE_IO_FILE_ACCESS_WRITE, file_contents->buffer, release_callback,
+      host_allocator, out_file_handle);
+  if (!iree_status_is_ok(status)) {
+    iree_file_contents_free(file_contents);
+  }
+  return status;
+}
+
+struct ParameterizeGlobalsPass
+    : public ParameterizeGlobalsBase<ParameterizeGlobalsPass> {
+  ParameterizeGlobalsPass(std::string archivePathStr,
+                          std::string parameterNamespaceStr) {
+    this->archivePath = archivePathStr;
+    this->parameterNamespace = parameterNamespaceStr;
+  };
+  ParameterizeGlobalsPass(const ParameterizeGlobalsPass &pass)
+      : ParameterizeGlobalsPass(pass.archivePath, pass.parameterNamespace) {}
+
+  void runOnOperation() override {
+    // Nothing to do if no path specified.
+    if (archivePath.empty()) {
+      return;
+    }
+
+    MLIRContext *context = &getContext();
+    ModuleOp moduleOp = getOperation();
+
+    iree_allocator_t host_allocator = iree_allocator_system();
+
+    // Create the parameter archive builder.
+    iree_io_parameter_archive_builder_t builder;
+    iree_io_parameter_archive_builder_initialize(host_allocator, &builder);
+
+    auto deinitializeExit = llvm::make_scope_exit([&]() {
+      return iree_io_parameter_archive_builder_deinitialize(&builder);
+    });
+
+    SmallVector<IREE::Util::GlobalOp> constantGlobals;
+    // Walk the globals in the module.
+    for (auto global : moduleOp.getOps<IREE::Util::GlobalOp>()) {
+      // Mutable globals cannot become parameters.
+      if (global.getIsMutable()) {
+        continue;
+      }
+      // Only initialized globals can be parameterized, and we restrict to
+      // tensors.
+      auto initialValueAttr = global.getInitialValueAttr();
+      if (!initialValueAttr ||
+          !llvm::isa<RankedTensorType>(initialValueAttr.getType())) {
+        continue;
+      }
+
+      // The attribute must be serializable to be turned into a parameter.
+      auto serializableAttr =
+          dyn_cast<IREE::Util::SerializableAttrInterface>(initialValueAttr);
+      if (!serializableAttr) {
+        continue;
+      }
+
+      iree_io_physical_size_t storageSize = serializableAttr.getStorageSize();
+      StringRef name = global.getSymName();
+
+      // Add a data entry to the builder for this global.
+      iree_status_t status = iree_io_parameter_archive_builder_add_data_entry(
+          &builder,
+          iree_string_view_t{name.data(),
+                             static_cast<iree_host_size_t>(name.size())},
+          /*metadata=*/iree_const_byte_span_empty(),
+          /*alignment=*/IREE_IO_PARAMETER_ARCHIVE_DEFAULT_DATA_ALIGNMENT,
+          storageSize);
+      if (!iree_status_is_ok(status)) {
```

Comment:
iree_status_t must be released (iree_status_ignore/iree_status_free/etc) and we want the nice errors it has

here's a helper that does this:
https://github.com/openxla/iree/blob/6735c8a619404e7a8c4d12fd10192ee758482797/compiler/src/iree/compiler/ConstEval/Runtime.cpp#L22-L36

for now I'd just copy it in here but at some point if we grow more parts in the compiler that use runtime stuff we can move it to a helper (compiler/Utils/RuntimeUtils.h or something)